api/rest/restapi.go

```diff
@@ -958,6 +988,39 @@ func (api *API) parseCidOrError(w http.ResponseWriter, r *http.Request) types.Pi
 	return pin
 }
 
+func (api *API) parsePathOrError(w http.ResponseWriter, r *http.Request) types.PinSerial {
```

Comment:
This logic should be in cluster main component (which needs to provide `PinPath()` function too.

Other than that https://godoc.org/github.com/ipfs/go-path provides utilities to parse the path. Not all paths need to be resolved:

* Invalid paths can error directly
* `IsJustAKey` paths can be pinned/unpinned directly.